### PR TITLE
pageserver: cancellation for remote ops in tenant deletion on shutdown

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

Tenant deletion had a couple of TODOs where we weren't using proper cancellation tokens that would have aborted the deletions during process shutdown.

## Summary of changes

- Refactor enough that deletion/shutdown code has access to the TenantManager's cancellation toke
- Use that cancellation token in tenant deletion instead of dummy tokens.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/task_mgr.rs [hunk 1]
```diff
@@ -453,7 +451,7 @@ async fn task_finish(
     }
 
     if shutdown_process {
-        shutdown_pageserver(None, 1).await;
+        std::process::exit(1);
     }
```
Comment (problame) [inline pageserver/src/task_mgr.rs:455]: I suppose this fixes https://github.com/neondatabase/neon/issues/3402 ?

File: pageserver/src/tenant/mgr.rs [hunk 2]
```diff
@@ -261,6 +261,10 @@ pub struct TenantManager {
     // See https://github.com/neondatabase/neon/issues/5796
     tenants: &'static std::sync::RwLock<TenantsMap>,
     resources: TenantSharedResources,
+
+    // This token is for use in edge cases like tenant deletion, where we are doing work on a
+    // tenant after it has been shut down, during deletion.
+    cancel: CancellationToken,
 }
```
Comment (problame) [inline pageserver/src/tenant/mgr.rs:268]: The description sounds tenant-scoped, why is it in TenantManager?

It'll probably evident when reading the rest of the PR, but, on first sight, it feels like this is missing some context.


```suggestion

    /// Long-running activities that operate on the tenant namespace instead of "inside" a Tenant
    /// should be sensitive to this token.
    ///
    /// Cancelled on [`TenantManager::shutdown`]
    cancel: CancellationToken,
}
```

File: pageserver/src/tenant/delete.rs [hunk 3]
```diff
@@ -546,8 +544,7 @@ impl DeleteTenantFlow {
             conf,
             remote_storage.as_ref(),
             &tenant.tenant_shard_id,
-            // Can't use tenant.cancel, it's already shut down.  TODO: wire in an appropriate token
-            &CancellationToken::new(),
+            &task_mgr::shutdown_token(),
```
Comment (problame) [inline pageserver/src/tenant/delete.rs:547]: This is a step forward, but, why not go through the trouble of plumbing through the `CancellationToken` from the two callers in this PR?

Comment (problame) [reply pageserver/src/tenant/delete.rs:547]: I.e., 
image

**Review by problame — APPROVED** [review]

File: pageserver/src/task_mgr.rs (hunk 1 above)
Comment (jcsp) [reply pageserver/src/task_mgr.rs:455]: Good find -- I had no idea that ticket existed.  Yes, I've just closed it.

File: pageserver/src/tenant/mgr.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:268]: Yes, this comment wasn't very well worded.

Rewritten to:
```
    // Long-running operations that happen outside of a [`Tenant`] lifetime should respect this token.
    // This is for edge cases like tenant deletion.  In normal cases (within a Tenant lifetime),
    // tenants have their own cancellation tokens, which we fire individually in [`Self::shutdown`], or
    // when the tenant detaches.
```

File: pageserver/src/tenant/delete.rs (hunk 3 above)
Comment (jcsp) [reply pageserver/src/tenant/delete.rs:547]: It's a bit awkward:
- If I provide a `CancellationToken` argument, callers have to be careful not to pass in the tenant's cancellation token (rather counter-intuitively) -- they need the TenantManager-scoped one, or a task scope one.
- Getting at such a token from the `spawn`->`resume_from_attach` path means passing it into `spawn`, which is also kind of confusing to read -- the argument has this special property of "please only use this cancellation token for deletions" (e.g. it would be wrong to use it as the parent for the cancellation token for the tenant being spawned).

So TLDR it's doable, but would probably come with a lot of comment-lines to explain it all at each step in the call stack.

The end state for this stuff should be simplified deletion, where we get rid of this horrible complexity that comes from sometimes spawning a deletion from the tenant load process:
- #5080
